app/assets/javascripts/discourse/app/static/prosemirror/components/prosemirror-editor.gjs

```diff
@@ -0,0 +1,253 @@
+
+  @action
+  handleAsyncPlugin(plugin) {
+    const state = this.view.state.reconfigure({
+      plugins: [...this.view.state.plugins, plugin],
+    });
+
+    this.view.updateState(state);
+  }
+
+  @action
+  setup(container) {
+    const params = this.pluginParams;
+
+    const plugins = [
+      buildInputRules(this.extensions, this.schema, this.args.includeDefault),
+      keymap(
+        buildKeymap(
+          this.extensions,
+          this.schema,
+          this.keymapFromArgs,
+          params,
+          this.args.includeDefault
+        )
+      ),
+      keymap(baseKeymap),
+      dropCursor({ color: "var(--primary)" }),
+      gapCursor(),
+      history(),
+      ...extractPlugins(this.extensions, params, this.handleAsyncPlugin),
+    ];
+
+    this.parser = new Parser(this.extensions, this.args.includeDefault);
+    this.serializer = new Serializer(this.extensions, this.args.includeDefault);
+
+    const state = EditorState.create({ schema: this.schema, plugins });
+
+    this.view = new EditorView(container, {
+      state,
+      nodeViews: extractNodeViews(this.extensions),
+      attributes: { class: this.args.class },
+      editable: () => this.args.disabled !== true,
+      dispatchTransaction: (tr) => {
+        this.view.updateState(this.view.state.apply(tr));
+
+        if (tr.docChanged && tr.getMeta("addToHistory") !== false) {
+          // If this gets expensive, we can debounce it
+          const value = this.serializer.convert(this.view.state.doc);
+          this.#lastSerialized = value;
+          this.args.change?.({ target: { value } });
+        }
+      },
+      handleDOMEvents: {
+        focus: () => {
+          this.args.focusIn?.();
+          return false;
+        },
+        blur: () => {
+          next(() => this.args.focusOut?.());
+          return false;
+        },
+      },
+      handleKeyDown: (view, event) => {
+        // suppress if Enter/Tab and the autocomplete is open
+        return (
+          AUTOCOMPLETE_KEY_DOWN_SUPPRESS.includes(event.key) &&
+          !!document.querySelector(".autocomplete")
+        );
+      },
+    });
+
+    this.textManipulation = new TextManipulation(getOwner(this), {
+      schema: this.schema,
+      view: this.view,
+      convertFromMarkdown: this.convertFromMarkdown,
+      convertToMarkdown: this.serializer.convert.bind(this.serializer),
+    });
+
+    this.#destructor = this.args.onSetup?.(this.textManipulation);
+
+    this.convertFromValue();
+  }
+
+  @bind
+  convertFromMarkdown(markdown) {
+    let doc;
+    try {
+      doc = this.parser.convert(this.schema, markdown);
+    } catch (e) {
+      next(() => this.dialog.alert(e.message));
+      throw e;
+    }
+    return doc;
+  }
+
+  @bind
+  convertFromValue() {
+    // Ignore the markdown we just serialized
+    if (this.args.value === this.#lastSerialized) {
+      return;
+    }
+
+    const doc = this.convertFromMarkdown(this.args.value);
+
+    const tr = this.view.state.tr;
```

Comment:
can we get a better name than tr ?

Reply:
Using `tr` when creating [a new transaction object](https://prosemirror.net/docs/ref/#state.EditorState.tr) from the state is kind of a standard within the ProseMirror ecosystem, though I'm not opposed to an alternative naming.